// @serenity-js/serenity-bdd/lib/stage/crew/serenity-bdd-reporter/processors/single-scene/SingleSceneEventQueueProcessor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SingleSceneEventQueueProcessor = void 0;
const events_1 = require("@serenity-js/core/lib/events");
const tiny_types_1 = require("tiny-types");
const EventQueueProcessor_1 = require("../EventQueueProcessor");
const transformations_1 = require("../transformations");
const SingleSceneReportContext_1 = require("./SingleSceneReportContext");
/**
 * @package
 */
class SingleSceneEventQueueProcessor extends EventQueueProcessor_1.EventQueueProcessor {
    supports(queue) {
        return queue
            && queue.first() instanceof events_1.SceneStarts;
    }
    process(queue) {
        return queue.reduce((context, event) => (0, tiny_types_1.match)(event)
            .when(events_1.SceneStarts, this.onSceneStarts(context))
            .when(events_1.FeatureNarrativeDetected, this.onFeatureNarrativeDetected(context))
            .when(events_1.SceneBackgroundDetected, this.onSceneBackgroundDetected(context))
            .when(events_1.SceneDescriptionDetected, this.onSceneDescriptionDetected(context))
            .when(events_1.BusinessRuleDetected, this.onBusinessRuleDetected(context))
            .when(events_1.TestRunnerDetected, this.onTestRunnerDetected(context))
            .when(events_1.SceneTagged, this.onSceneTagged(context))
            .when(events_1.ActivityStarts, this.onActivityStarts(context))
            .when(events_1.ActivityFinished, this.onActivityFinished(context))
            .when(events_1.ActivityRelatedArtifactGenerated, this.onActivityRelatedArtifactGenerated(context))
            .when(events_1.ActivityRelatedArtifactArchived, this.onActivityRelatedArtifactArchived(context))
            .when(events_1.SceneFinished, this.onSceneFinished(context))
            .else(() => context), new SingleSceneReportContext_1.SingleSceneReportContext() // eslint-disable-line @typescript-eslint/indent
        ).build();
    }
    onSceneStarts(report) {
        return (event) => report
            .with((0, transformations_1.reportIdIncluding)(event.details.category.value, event.details.name.value))
            .with((0, transformations_1.scenarioDetailsOf)(event.details))
            .with((0, transformations_1.executionStartedAt)(event.timestamp));
    }
    onActivityStarts(report) {
        return (event) => report
            .with((0, transformations_1.activityStarted)(event.activityId, event.details.name, event.timestamp));
    }
    onActivityFinished(report) {
        return (event) => report
            .with((0, transformations_1.activityFinished)(event.activityId, event.outcome, event.timestamp));
    }
    onSceneFinished(report) {
        return (event) => report
            .with((0, transformations_1.executionFinishedWith)(event.outcome))
            .with((0, transformations_1.executionFinishedAt)(event.timestamp));
    }
}
exports.SingleSceneEventQueueProcessor = SingleSceneEventQueueProcessor;
