// @serenity-js/serenity-bdd/lib/stage/crew/serenity-bdd-reporter/processors/EventQueueProcessors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventQueueProcessors = void 0;
const events_1 = require("@serenity-js/core/lib/events");
const model_1 = require("@serenity-js/core/lib/model");
const scene_sequence_1 = require("./scene-sequence");
const single_scene_1 = require("./single-scene");
/**
 * @package
 */
class EventQueueProcessors {
    constructor() {
        this.singleSceneProcessor = new single_scene_1.SingleSceneEventQueueProcessor();
        this.sceneSequenceProcessor = new scene_sequence_1.SceneSequenceEventQueueProcessor();
    }
    // todo: move `name` to Artifact and return Artifact[]... and sceneId?
    process(queues) {
        const results = [];
        queues.forEach(queue => {
            const report = queue.first() instanceof events_1.SceneStarts
                ? this.singleSceneProcessor.process(queue)
                : this.sceneSequenceProcessor.process(queue);
            results.push({
                name: new model_1.Name(report.name),
                artifact: model_1.TestReport.fromJSON(report),
                sceneId: queue.sceneId,
            });
        });
        return results;
    }
}
exports.EventQueueProcessors = EventQueueProcessors;
