// @serenity-js/serenity-bdd/lib/stage/crew/serenity-bdd-reporter/processors/EventQueueProcessor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventQueueProcessor = void 0;
const transformations_1 = require("./transformations");
/**
 * @package
 */
class EventQueueProcessor {
    onFeatureNarrativeDetected(report) {
        return (event) => report
            .with((0, transformations_1.featureNarrativeOf)(event.description));
    }
    onSceneBackgroundDetected(report) {
        return (event) => report
            .with((0, transformations_1.backgroundOf)(event.name, event.description));
    }
    onSceneDescriptionDetected(report) {
        return (event) => report
            .with((0, transformations_1.descriptionOf)(event.description));
    }
    onTestRunnerDetected(report) {
        return (event) => report
            .with((0, transformations_1.testRunnerCalled)(event.name));
    }
    onSceneTagged(report) {
        return (event) => report
            .with((0, transformations_1.tagOf)(event.tag));
    }
    onBusinessRuleDetected(report) {
        return (event) => report
            .with((0, transformations_1.businessRuleOf)(event.rule));
    }
    onActivityRelatedArtifactGenerated(report) {
        return (event) => report
            .with((0, transformations_1.activityRelatedArtifact)(event.activityId, event.name, event.artifact, event.timestamp));
    }
    onActivityRelatedArtifactArchived(report) {
        return (event) => report
            .with((0, transformations_1.archivedActivityRelatedArtifact)(event.activityId, event.type, event.path, event.timestamp));
    }
}
exports.EventQueueProcessor = EventQueueProcessor;
