// @serenity-js/console-reporter/lib/stage/crew/console-reporter/Printer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Printer = void 0;
/**
 * @desc
 *  A simple abstraction around the standard output stream
 *  with support for indenting the printed text.
 *
 * @public
 */
class Printer {
    constructor(stdout = process.stdout) {
        this.stdout = stdout;
        this.indentation = new Indentation(2);
    }
    /**
     * @desc
     *  Prints the `args` to the `stdout`, followed by a new line character (`\n`).
     *
     *  Please note that this method bypasses any indentation as it's intended
     *  to be used to print separator lines, etc.
     *
     * @param {...any[]} args
     *
     * @returns {void}
     */
    println(...args) {
        return this.print(...args, '\n');
    }
    /**
     * @desc
     *  Prints the `args` to the `stdout`, indenting the output according to
     *  the current indentation level.
     *
     * @param {...any[]} args
     *
     * @returns {void}
     */
    print(...args) {
        this.stdout.write(this.indentation.indented(`${args.map(String).join('')}`));
    }
    /**
     * @desc
     *  Increases the current indentation by 2 spaces.
     *
     * @returns {void}
     */
    indent() {
        this.indentation.increase();
    }
    /**
     * @desc
     *  Reduces the current indentation by 2 spaces.
     *
     * @returns {void}
     */
    outdent() {
        this.indentation.decrease();
    }
}
exports.Printer = Printer;
/**
 * @package
 */
class Indentation {
    constructor(step) {
        this.step = step;
        this.current = 0;
    }
    increase() {
        this.current += this.step;
    }
    decrease() {
        if (this.current - this.step >= 0) {
            this.current -= this.step;
        }
    }
    indented(...fragments) {
        return fragments.join('')
            .split('\n')
            .map(line => line
            ? ' '.repeat(this.current) + line
            : '')
            .join('\n');
    }
}
